File: 10-structs/main.zig
const std = @import("std");

// === 基本结构体 ===
const Point = struct {
    x: f64,
    y: f64,

    // 方法：显式 self 参数
    pub fn distanceTo(self: Point, other: Point) f64 {
        const dx = self.x - other.x;
        const dy = self.y - other.y;
        return @sqrt(dx * dx + dy * dy);
    }

    // 构造函数惯用法
    pub fn init(x: f64, y: f64) Point {
        return .{ .x = x, .y = y };
    }

    // 自定义格式化（Zig 0.15 新签名）
    pub fn format(self: Point, writer: anytype) !void {
        try writer.print("Point({d:.1}, {d:.1})", .{ self.x, self.y });
    }
};

// === 带默认值的结构体 ===
const Config = struct {
    host: []const u8 = "localhost",
    port: u16 = 8080,
    debug: bool = false,
};

// === packed struct：精确控制内存布局 ===
const Flags = packed struct {
    read: bool,
    write: bool,
    execute: bool,
    _reserved: u5 = 0,
};

// === extern struct：C 兼容布局 ===
const CPoint = extern struct {
    x: f64,
    y: f64,
};

// === 带指针参数的方法（可变 self）===
const Counter = struct {
    count: u32 = 0,

    pub fn increment(self: *Counter) void {
        self.count += 1;
    }

    pub fn getCount(self: Counter) u32 {
        return self.count;
    }
};

// === 泛型结构体 ===
fn Pair(comptime T: type) type {
    return struct {
        first: T,
        second: T,

        const Self = @This();

        pub fn init(first: T, second: T) Self {
            return .{ .first = first, .second = second };
        }

        pub fn swap(self: Self) Self {
            return .{ .first = self.second, .second = self.first };
        }
    };
}

pub fn main() !void {
    var buf: [4096]u8 = undefined;
    const file = std.fs.File{ .handle = std.posix.STDOUT_FILENO };
    var w = file.writer(&buf);
    defer w.interface.flush() catch {};

    // === 基本使用 ===
    const p1 = Point.init(0, 0);
    const p2 = Point{ .x = 3, .y = 4 };
    try w.interface.print("{f} 到 {f} 的距离: {d:.2}\n", .{ p1, p2, p1.distanceTo(p2) });

    // === 默认值 ===
    const default_cfg = Config{};
    const custom_cfg = Config{ .port = 9090, .debug = true };
    try w.interface.print("默认配置: {s}:{}\n", .{ default_cfg.host, default_cfg.port });
    try w.interface.print("自定义配置: {s}:{} debug={}\n", .{ custom_cfg.host, custom_cfg.port, custom_cfg.debug });

    // === packed struct ===
    const flags = Flags{ .read = true, .write = true, .execute = false };
    const as_byte: u8 = @bitCast(flags);
    try w.interface.print("Flags 位表示: 0b{b:0>8}\n", .{as_byte});
    try w.interface.print("Flags 大小: {} 字节\n", .{@sizeOf(Flags)});

    // === 可变方法 ===
    var counter = Counter{};
    counter.increment();
    counter.increment();
    counter.increment();
    try w.interface.print("计数器: {}\n", .{counter.getCount()});

    // === 匿名结构体 ===
    const anon = .{ .name = "Zig", .version = @as(u32, 13) };
    try w.interface.print("匿名结构体: name={s}, version={}\n", .{ anon.name, anon.version });

    // === 元组 ===
    const tuple = .{ @as(i32, 42), "hello", true };
    try w.interface.print("元组: {}, {s}, {}\n", .{ tuple[0], tuple[1], tuple[2] });

    // === 泛型结构体 ===
    const int_pair = Pair(i32).init(10, 20);
    const swapped = int_pair.swap();
    try w.interface.print("原始: ({}, {})\n", .{ int_pair.first, int_pair.second });
    try w.interface.print("交换: ({}, {})\n", .{ swapped.first, swapped.second });

    // === 结构体大小 ===
    try w.interface.print("\n--- 内存布局 ---\n", .{});
    try w.interface.print("Point 大小: {} 字节\n", .{@sizeOf(Point)});
    try w.interface.print("CPoint 大小: {} 字节\n", .{@sizeOf(CPoint)});
    try w.interface.print("Config 大小: {} 字节\n", .{@sizeOf(Config)});
}

File: 10-structs/main.rs
use std::fmt;
use std::mem;

// === 基本结构体 ===
struct Point {
    x: f64,
    y: f64,
}

// impl 块定义方法
impl Point {
    // 关联函数（类似构造函数）
    fn new(x: f64, y: f64) -> Self {
        Point { x, y }
    }

    // 方法：&self 借用
    fn distance_to(&self, other: &Point) -> f64 {
        let dx = self.x - other.x;
        let dy = self.y - other.y;
        (dx * dx + dy * dy).sqrt()
    }
}

impl fmt::Display for Point {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "Point({:.1}, {:.1})", self.x, self.y)
    }
}

// === 带 Default trait 的结构体 ===
struct Config {
    host: String,
    port: u16,
    debug: bool,
}

impl Default for Config {
    fn default() -> Self {
        Config {
            host: String::from("localhost"),
            port: 8080,
            debug: false,
        }
    }
}

impl fmt::Display for Config {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "Config{{ host={}, port={}, debug={} }}", self.host, self.port, self.debug)
    }
}

// === 可变方法 ===
struct Counter {
    count: u32,
}

impl Counter {
    fn new() -> Self {
        Counter { count: 0 }
    }

    fn increment(&mut self) {
        self.count += 1;
    }

    fn get_count(&self) -> u32 {
        self.count
    }
}

// === 元组结构体 ===
struct Color(u8, u8, u8);

impl fmt::Display for Color {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "Color({}, {}, {})", self.0, self.1, self.2)
    }
}

// === 单元结构体 ===
struct Unit;

// === 泛型结构体 ===
#[derive(Debug)]
struct Pair<T> {
    first: T,
    second: T,
}

impl<T: Clone> Pair<T> {
    fn new(first: T, second: T) -> Self {
        Pair { first, second }
    }

    fn swap(&self) -> Self {
        Pair {
            first: self.second.clone(),
            second: self.first.clone(),
        }
    }
}

// === C 兼容布局 ===
#[repr(C)]
struct CPoint {
    x: f64,
    y: f64,
}

// === packed 布局 ===
#[repr(packed)]
struct PackedData {
    a: u8,
    b: u32,
    c: u8,
}

fn main() {
    // === 基本使用 ===
    let p1 = Point::new(0.0, 0.0);
    let p2 = Point { x: 3.0, y: 4.0 };
    println!("{} 到 {} 的距离: {:.2}", p1, p2, p1.distance_to(&p2));

    // === Default trait ===
    let default_cfg = Config::default();
    let custom_cfg = Config {
        port: 9090,
        debug: true,
        ..Config::default() // 用 default 填充其余字段
    };
    println!("默认配置: {}", default_cfg);
    println!("自定义配置: {}", custom_cfg);

    // === 可变方法 ===
    let mut counter = Counter::new();
    counter.increment();
    counter.increment();
    counter.increment();
    println!("计数器: {}", counter.get_count());

    // === 元组结构体 ===
    let red = Color(255, 0, 0);
    println!("颜色: {}", red);

    // === 解构 ===
    let Point { x, y } = Point::new(10.0, 20.0);
    println!("解构: x={}, y={}", x, y);

    // === 泛型结构体 ===
    let int_pair = Pair::new(10, 20);
    let swapped = int_pair.swap();
    println!("原始: ({}, {})", int_pair.first, int_pair.second);
    println!("交换: ({}, {})", swapped.first, swapped.second);

    // === 所有权与方法 ===
    // &self: 借用，调用后仍可使用
    // &mut self: 可变借用
    // self: 消耗所有权，调用后不可再使用

    // === 结构体大小 ===
    println!("\n--- 内存布局 ---");
    println!("Point 大小: {} 字节", mem::size_of::<Point>());
    println!("CPoint 大小: {} 字节", mem::size_of::<CPoint>());
    println!("Config 大小: {} 字节", mem::size_of::<Config>());
    println!("PackedData 大小: {} 字节 (未 packed 时可能更大)", mem::size_of::<PackedData>());
    println!("Unit 大小: {} 字节", mem::size_of::<Unit>());
}
